### PR TITLE
[ppi]: Implement port bulk comparison logic

### PR DESCRIPTION
**Pull request by nazariig**
Signed-off-by: Nazarii Hnydyn <nazariig@nvidia.com>

**DEPENDS:**
1. https://github.com/sonic-net/sonic-sairedis/pull/1171
2. https://github.com/sonic-net/sonic-sairedis/pull/1197
3. https://github.com/sonic-net/sonic-swss/pull/2806

**HLD:** https://github.com/sonic-net/SONiC/pull/1084

**What I did**
* Implemented port bulk add/remove comparison logic
* Refactored Port OA to facilitate port config operations

**Why I did it**
* To improve switch Fast Boot startup time

**How I verified it**
* UT tests
* VS tests

**Details if related**
* N/A

### CODE REVIEW

**Comment by prgeor** [comment]
@nazariig can you attach the test results to the PR?

File: orchagent/portsorch.cpp [hunk 1]
```diff
@@ -448,191 +466,455 @@ PortsOrch::PortsOrch(DBConnector *db, DBConnector *stateDb, vector<table_name_wi
...
+{
+    return this->m_portConfigState;
+}
+
+void PortsOrch::setPortConfigState(port_config_state_t value)
+{
+    this->m_portConfigState = value;
+}
+
+bool PortsOrch::addPortBulk(const std::vector<PortConfig> &portList)
```
Comment (prgeor) [inline orchagent/portsorch.cpp:683]: Please add comment what this fn is doing

**Comment by prgeor** [comment]
@nazariig can you provide more detailed test results. Have you tested AN/LT? FEC?

File: orchagent/portsorch.cpp (hunk 1 above)
Comment (nazariig) [reply orchagent/portsorch.cpp:683]: @prgeor done

**Comment by nazariig** [comment]
> @nazariig can you provide more detailed test results. Have you tested AN/LT? FEC?

@prgeor done. All VS tests are passed. BTW, we don't have a way to test LT. Maybe you can help with that?

**Comment by prgeor** [comment]
@nazariig please share the performance data for ports create as part of create_switch (SAI profile file) vs bulk ports create from portsorch for your platform

**Comment by prgeor** [comment]
> > @nazariig can you provide more detailed test results. Have you tested AN/LT? FEC?
> 
> @prgeor done. All VS tests are passed. BTW, we don't have a way to test LT. Maybe you can help with that?

@nazariig can you test AN?

**Comment by prgeor** [comment]
> @nazariig can you provide more detailed test results. Have you tested AN/LT? FEC?

@nazariig can you test setting of FEC?

File: orchagent/portsorch.cpp [hunk 2]
```diff
@@ -448,191 +466,460 @@ PortsOrch::PortsOrch(DBConnector *db, DBConnector *stateDb, vector<table_name_wi
...
+            attr.value.s32 = cit.pfc_asym.value;
+            attrList.push_back(attr);
+        }
+
+        attrDataList.push_back(attrList);
+        attrCountList.push_back(static_cast<std::uint32_t>(attrDataList.back().size()));
+        attrPtrList.push_back(attrDataList.back().data());
+    }
+
+    auto status = sai_port_api->create_ports(
```
Comment (prgeor) [inline orchagent/portsorch.cpp:808]: From HLD, what if create_ports is NULL or not supported?

image

**Comment by prgeor** [comment]
@nazariig can you simulate a scenario where your changes are tested on your platform where it does not support bulk port creation?

File: orchagent/portsorch.cpp (hunk 2 above)
Comment (nazariig) [reply orchagent/portsorch.cpp:808]: @prgeor these statements are not relevant:
1. Checking for NULL doesn't make any sense - the bulk API (sairedis) is implemented and currently available for all the vendors
2. The current syncd implementation doesn't propagate `SAI_STATUS_NOT_SUPPORTED` and `SAI_STATUS_NOT_IMPLEMENTED` codes to upper layer. Instead, it automatically falls back to the legacy one-by-one processing of the requested bulk operation objects. Taking this into consideration, there is no point to have a capabilities check in swss

**Comment by nazariig** [comment]
> @nazariig please share the performance data for ports create as part of create_switch (SAI profile file) vs bulk ports create from portsorch for your platform

it is not relevant in context of this feature, since `create_switch` part hasn't been touched.

**Comment by nazariig** [comment]
> > > @nazariig can you provide more detailed test results. Have you tested AN/LT? FEC?
> > 
> > 
> > @prgeor done. All VS tests are passed. BTW, we don't have a way to test LT. Maybe you can help with that?
> 
> @nazariig can you test AN?

yes, we can

**Comment by nazariig** [comment]
> > @nazariig can you provide more detailed test results. Have you tested AN/LT? FEC?
> 
> @nazariig can you test setting of FEC?

@prgeor yes, we can

**Comment by prgeor** [comment]
@liat-grozovik  @nazariig there is merge conflict

**Comment by zhangyanzhao** [comment]
@nazariig @prgeor  any open issue still left?

**Comment by prgeor** [comment]
@nazariig please take care of merge conflict.

**Comment by nazariig** [comment]
@zhangyanzhao yes. We need to do some additional testing and then get back to @prgeor

**Comment by nazariig** [comment]
VS tests failed:
```
test_evpn_fdb.py::test_evpnFdb xargs: sudo: terminated by signal 9
```

File: orchagent/portsorch.cpp [hunk 3]
```diff
@@ -638,6 +567,312 @@ PortsOrch::PortsOrch(DBConnector *db, DBConnector *stateDb, vector<table_name_wi
...
+{
+    this->m_portConfigState = value;
+}
+
+bool PortsOrch::addPortBulk(const std::vector<PortConfig> &portList)
+{
+    // The method is used to create ports in a bulk mode.
+    // The action takes place when:
+    // 1. Ports are being initialized at system start
+    // 2. Ports are being added/removed by a user at runtime
```
Comment (prgeor) [inline orchagent/portsorch.cpp:689]: @nazariig #2 is done via breakout CLI? What other way user can use?

**Comment by zhangyanzhao** [comment]
Please finish the review.

File: orchagent/portsorch.cpp (hunk 3 above)
Comment (nazariig) [reply orchagent/portsorch.cpp:689]: @prgeor this means that api will be invoked each time there is an attempt to modify port related stuff in db. Breakout CLI does exactly the same

**Comment by nazariig** [comment]
@prgeor do you have other comments regarding the implementation?

**Comment by zhangyanzhao** [comment]
@nazariig  can you please check and fix the build failure?

**Comment by prgeor** [comment]
> @prgeor do you have other comments regarding the implementation?

@nazariig need to re-review since the last changes.

**Comment by liat-grozovik** [comment]
@prgeor @saiarcot895 could you please help to review/merge?

**Comment by xumia** [comment]
@nazariig , could you please try the command "/azpw run Azure.sonic-swss" again?

**Comment by nazariig** [comment]
@xumia it works! Did you fix the issue?

**Comment by nazariig** [comment]
> @nazariig , could you please try the command "/azpw run Azure.sonic-swss" again?

@xumia it works! Did you fix the issue?

**Comment by liat-grozovik** [comment]
@prgeor @lguohan please be reminded this is critical feature for 202305 and 202211. please close the review ASAP.

**Review by liat-grozovik — APPROVED** [review]

**Comment by liat-grozovik** [comment]
Backported PR was approved and merged. moving forward with this PR as well.

**Pull request merged**